File: DCP-1829/solution.swift
// Problem 1829
// Strategy Move of always winning in a ghost game

import Foundation

class TrieNode {
    var children: [Character: TrieNode] = [:]
    var isEndOfWord: Bool = false
}

func optimalStartingLetters(dictionary: [String]) -> [Character] {
    let root = TrieNode()
    
    // Build Trie
    for word in dictionary {
        var node = root
        for letter in word {
            if node.children[letter] == nil {
                node.children[letter] = TrieNode()
            }
            node = node.children[letter]!
        }
        node.isEndOfWord = true
    }
    
    func canWin(node: TrieNode, isPlayer1Turn: Bool) -> Bool {
        if node.isEndOfWord {
            return !isPlayer1Turn
        }
        
        for (_, childNode) in node.children {
            if !canWin(node: childNode, isPlayer1Turn: !isPlayer1Turn) {
                return true
            }
        }
        
        return false
    }
    
    var optimalStarts: [Character] = []
    
    for (letter, node) in root.children {
        if canWin(node: node, isPlayer1Turn: false) {
            optimalStarts.append(letter)
        }
    }
    
    return optimalStarts
}

// Example usage
let dictionary = ["cat", "calf", "dog", "bear"]
let result = optimalStartingLetters(dictionary: dictionary)
print(result) // Output: ["b", "c", "d"]

File: DCP-1829/solution.kt
class TrieNode {
    val children = mutableMapOf<Char, TrieNode>()
    var isEndOfWord = false
}

fun optimalStartingLetters(dictionary: List<String>): List<Char> {
    val root = TrieNode()
    
    // Build Trie
    for (word in dictionary) {
        var node = root
        for (letter in word) {
            node.children.putIfAbsent(letter, TrieNode()).also { node = it }
        }
        node.isEndOfWord = true
    }
    
    fun canWin(node: TrieNode, isPlayer1Turn: Boolean): Boolean {
        if (node.isEndOfWord) return !isPlayer1Turn
        
        for (childNode in node.children.values) {
            if (!canWin(childNode, !isPlayer1Turn)) return true
        }
        
        return false
    }

    val optimalStarts = mutableListOf<Char>()
    
    for ((letter, node) in root.children) {
        if (canWin(node, false)) {
            optimalStarts.add(letter)
        }
    }
    
    return optimalStarts
}

// Example usage
val dictionary = listOf("cat", "calf", "dog", "bear")
val result = optimalStartingLetters(dictionary)
println(result) // Output: [b, c, d]